postgresql: Improve and cleanup module.

aszlig opened the pull request:
Essentially why I'm doing a pull request instead of directly pushing here is because of a6c2e79af979530769e0727bf51a735e3f3f1297, which could break configurations out there.

At the current master `enableTCPIP` has no effect at all, because of the reason I described in the mentioned commit.

So, with this commit, it will essentially break configurations where users have just ignored `enableTCPIP`, which leads to PostgreSQL listening to local TCP/IP connections by default (which I presume wasn't intended, because `enableTCPIP` defaults to `false`).

Which brings me to the main question: Should we enable listening to local TCP/IP connections by default?

aszlig committed:
postgresql: Use local peer auth for >= v9.1.

The new keyword for local authentication is "peer" and "ident" only works for
TCP/IP. However, ident is still accepted for local connections, but is currently
kept for backwards compatibility reasons, see here:

http://www.postgresql.org/docs/9.1/static/release-9-1.html#AEN108218

Details about peer auth and ident auth can be found here:

http://www.postgresql.org/docs/9.1/static/auth-methods.html#AUTH-IDENT
http://www.postgresql.org/docs/9.1/static/auth-methods.html#AUTH-PEER

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
postgresql: Remove unused authMethod option.

I think we can still introduce it later if needed, but in the end I think it
would make more sense to allow ovverriding the pg_hba config without using
extraConfig and adding hba_file.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
postgresql: Use listen_addresses instead of -i.

This replaces the enableTCPIP option with a new listenAddresses option. The
reason is because listen_addresses is by default set to "localhost", so omitting
-i isn't effective, especially in the more recent versions.

Have a look at the point on -i, which essentially deprecates the use of it (and
that since a long time, even for 8.4):

http://www.postgresql.org/docs/9.2/static/app-postgres.html#APP-POSTGRES-OPTIONS

Nevertheless, listen_addresses is the correct way to specify the addresses the
server should listen on. So let's use it and provide an option for users to
change it as necessary.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
postgresql: Allow to override pg_hba settings.

Until this commit, it was only possible to append to pg_hba.conf, so if you want
to specify mappings and other options, this wasn't possible.

Still, it's only possible to append, but we now restrict the basic configuration
to an absolute minimum, only allowing the root user to connect. That way if you
want to specify access rules for other users, existing auth rules won't get into
the way.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
@edolstra: Any objections? If not, I'll merge.

edolstra commented on modules/services/databases/postgresql.nix:
```diff
@@ -81,9 +92,15 @@ in
           Defines how users authenticate themselves to the server.
+          This is in the format of the pg_hba.conf, for documentation, see:
+
+          http://www.postgresql.org/docs/9.2/static/auth-pg-hba-conf.html
```
Note that descriptions are in DocBook format, so it's better to write something like:

```
This is in the format of 
<link xlink:href="http://www.postgresql.org/docs/9.2/static/auth-pg-hba-conf.html">the 
<filename>pg_hba.conf</filename> configuration file</link>.
```

or something like that. That way it shows up as an actual link in the NixOS manual.

aszlig replied:
Ah, thanks! Going to change that now... :-)

edolstra commented on modules/services/databases/postgresql.nix:
```diff
@@ -22,18 +22,29 @@ let
-  flags = optional cfg.enableTCPIP "-i";
+  authentication = ''
+    # Generated file; do not edit!
+    local all root ${localAuthMethod}
```
This changes the default for local connections from "local all all" to "local all root". That seems to lock out non-root users.

aszlig replied:
Hm, do you have a better (and possibly simpler) idea of solving this?

I actually could change the [default value](https://github.com/aszlig/nixos/blob/ebb2c53741cae02afd6fa3d7b4ec017f1479ae45/modules/services/databases/postgresql.nix#L96-L97) to include peer connections for all accounts. Although, it would bring up the problem that it depends on the specified [package](https://github.com/aszlig/nixos/blob/ebb2c53741cae02afd6fa3d7b4ec017f1479ae45/modules/services/databases/postgresql.nix#L66-L71), eventually rendering it into a circular dependency.

edolstra commented:
Instead of removing enableTCPIP (which breaks configurations that use it), it could be mapped to `listenAddresses = ["*"]`.

The default should be Unix domain sockets only.

aszlig committed:
postgresql: Add docbook links for URLs.

Thanks to @edolstra for pointing it out here:

https://github.com/NixOS/nixos/pull/188#discussion_r4704438

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
Well, for that it would be a good idea to mark it deprecated, because `enableTCPIP` set to `false` has **no effect** at all. That's why I removed the option in the first place, because people might be unaware that PostgreSQL is actually listening to TCP/IP, even though they disabled it using this option.

aszlig commented:
@edolstra: ping?

domenkozar commented:
@aszlig let's close this?
